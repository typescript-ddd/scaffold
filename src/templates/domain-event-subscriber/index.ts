export * from "./domain-event-subscriber";