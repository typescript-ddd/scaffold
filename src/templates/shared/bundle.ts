import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
