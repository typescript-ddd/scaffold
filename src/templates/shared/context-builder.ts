import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
