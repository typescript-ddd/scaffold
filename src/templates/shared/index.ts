export * from "./context-builder";
export * from "./generate-context";