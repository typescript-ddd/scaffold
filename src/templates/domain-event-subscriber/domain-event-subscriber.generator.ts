import { Project } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { DomainEventSubscriberTemplateValues } from "./domain-event-subscriber.types";


export const generateDomainEventSubscriber = (
  values: DomainEventSubscriberTemplateValues,
  context: GenerateContext
) => {
  const { eventName } = values;
  const eventType = strings.capitalize(eventName);
  const subscriberType = `${eventType}Subscriber`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${subscriberType} implements DomainEventSubscriber<${eventType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEventSubscriber"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain", "event"),
    namedImports: [eventType],
  })

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(subscriberType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(subscriberType)}.`,
  });

  classDeclaration
    .addMethod({
      name: "on",
      parameters: [{ name: "event", type: eventType }],
      returnType: "void",
    })
    .setBodyText((writer) => {
      writer.writeLine(`// TODO: Implement subscriber logic.`);
    });

  let output = sourceFile.getFullText();

  return output;
};
