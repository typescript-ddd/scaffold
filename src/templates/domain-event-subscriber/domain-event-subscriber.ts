import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateDomainEventSubscriberOptions = {
  eventName: string;
};

export const generateDomainEventSubscriber = (
  options: GenerateDomainEventSubscriberOptions,
  context: GenerateContext
) => {
  const { eventName } = options;
  const eventType = strings.capitalize(eventName);
  const subscriberType = `${eventType}Subscriber`;

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${subscriberType} implements DomainEventSubscriber<${eventType}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["DomainEventSubscriber"],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain", "event"),
    namedImports: [eventType],
  })

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(subscriberType);
  classDeclaration.addJsDoc({
    description: `Represents ${strings.describe(subscriberType)}.`,
  });

  classDeclaration
    .addMethod({
      name: "on",
      parameters: [{ name: "event", type: eventType }],
      returnType: "void",
    })
    .setBodyText((writer) => {
      writer.writeLine(`// TODO: Implement subscriber logic.`);
    });

  let output = sourceFile.getFullText();

  return output;
};
