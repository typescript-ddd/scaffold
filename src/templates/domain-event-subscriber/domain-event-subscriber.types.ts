import { TemplateStringOption } from "../shared/template";

export type DomainEventSubscriberTemplateValues = {
  eventName: string;
};

export type DomainEventSubscriberTemplateOptions = {
  eventName: TemplateStringOption;
};
