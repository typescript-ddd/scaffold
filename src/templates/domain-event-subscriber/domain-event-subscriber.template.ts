import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateDomainEventSubscriber } from "./domain-event-subscriber.generator";
import {
  DomainEventSubscriberTemplateOptions,
  DomainEventSubscriberTemplateValues,
} from "./domain-event-subscriber.types";

export class DomainEventSubscriberTemplate
  implements
    Template<
      DomainEventSubscriberTemplateValues,
      DomainEventSubscriberTemplateOptions
    >
{
  readonly name = "Domain Event Subscriber";
  readonly description = "Generates a domain event subscriber";
  readonly options: DomainEventSubscriberTemplateOptions = {
    eventName: {
      name: "eventName",
      label: "Event Name",
      description: "The name of the event",
      optionType: "string",
    },
  };
  readonly defaultValues = {
    eventName: "UserCreated",
  };
  generate(values: DomainEventSubscriberTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateDomainEventSubscriber(values, context, chunkName);
  }
}
