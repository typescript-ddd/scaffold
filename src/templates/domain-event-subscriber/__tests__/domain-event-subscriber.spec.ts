import { ContextBuilder } from "../../shared";
import { generateDomainEventSubscriber } from "../domain-event-subscriber";

describe("DomainEventSubscriber", () => {
    it("should generate domain event subscriber", () => {
        // Arrange
        const options = {
        eventName: "UserCreatedEvent",
        };
        const context = ContextBuilder.build("@src");
    
        // Act
        const output = generateDomainEventSubscriber(options, context);
    
        // Assert
        expect(output).toMatchSnapshot();
    });
});