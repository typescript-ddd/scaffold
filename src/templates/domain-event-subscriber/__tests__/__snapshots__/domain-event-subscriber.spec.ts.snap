// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainEventSubscriber should generate domain event subscriber 1`] = `
"import { DomainEventSubscriber } from "@typescript-ddd/core";
import { UserCreatedEvent } from "@src/domain/event";

/** Represents a user created event subscriber. */
export class UserCreatedEventSubscriber implements DomainEventSubscriber<UserCreatedEvent> {
    on(event: UserCreatedEvent): void {
        // TODO: Implement subscriber logic.
    }
}    
"
`;
